include/gridtools/stencil_composition/backend_cuda/launch_kernel.hpp

```diff
@@ -0,0 +1,232 @@
+         *  \param i_size the size of the computation area in i dimension
+         *  \param j_size the size of the computation area in j dimension
+         *  \param zblocks number of the blocks in k dimension
+         *  \param shared_memory_size delegated to cuda
+         */
+        template <class Extent,
+            int_t BlockSizeI,
+            int_t BlockSizeJ,
+            class Fun,
+            std::enable_if_t<!launch_kernel_impl_::is_empty_ij_extents<Extent>(), int> = 0>
+        GT_FORCE_INLINE void launch_kernel(
+            int_t i_size, int_t j_size, uint_t zblocks, Fun fun, size_t shared_memory_size = 0) {
+            GT_STATIC_ASSERT(is_extent<Extent>::value, GT_INTERNAL_ERROR);
+            GT_STATIC_ASSERT(Extent::iminus::value <= 0, GT_INTERNAL_ERROR);
+            GT_STATIC_ASSERT(Extent::iplus::value >= 0, GT_INTERNAL_ERROR);
+
+            GT_STATIC_ASSERT(std::is_trivially_copyable<Fun>::value, GT_INTERNAL_ERROR);
+
+            static constexpr auto halo_lines = Extent::jplus::value - Extent::jminus::value +
+                                               (Extent::iminus::value < 0 ? 1 : 0) + (Extent::iplus::value > 0 ? 1 : 0);
+            static const size_t num_threads = BlockSizeI * (BlockSizeJ + halo_lines);
+
+            uint_t xblocks = (i_size + BlockSizeI - 1) / BlockSizeI;
+            uint_t yblocks = (j_size + BlockSizeJ - 1) / BlockSizeJ;
+
+            dim3 blocks = {xblocks, yblocks, zblocks};
+            dim3 threads = {BlockSizeI, BlockSizeJ + halo_lines, 1};
+
+            launch_kernel_impl_::wrapper<num_threads, BlockSizeI, BlockSizeJ, Extent>
+                <<<blocks, threads, shared_memory_size>>>(std::move(fun), i_size, j_size);
+
+#ifndef NDEBUG
+            GT_CUDA_CHECK(cudaDeviceSynchronize());
+#else
+            GT_CUDA_CHECK(cudaGetLastError());
+#endif
+        }
+
+        template <class Extent,
+            int_t BlockSizeI,
+            int_t BlockSizeJ,
+            class Fun,
+            std::enable_if_t<launch_kernel_impl_::is_empty_ij_extents<Extent>(), int> = 0>
+        GT_FORCE_INLINE void launch_kernel(
+            int_t i_size, int_t j_size, uint_t zblocks, Fun fun, size_t shared_memory_size = 0) {
+
+            GT_STATIC_ASSERT(std::is_trivially_copyable<Fun>::value, GT_INTERNAL_ERROR);
+
+            static const size_t num_threads = BlockSizeI * BlockSizeJ;
+
+            uint_t xblocks = (i_size + BlockSizeI - 1) / BlockSizeI;
+            uint_t yblocks = (j_size + BlockSizeJ - 1) / BlockSizeJ;
+
+            dim3 blocks = {xblocks, yblocks, zblocks};
+            dim3 threads = {BlockSizeI, BlockSizeJ, 1};
+
+            if (i_size % BlockSizeI == 0 && j_size % BlockSizeJ == 0)
+                launch_kernel_impl_::trivial_wrapper<num_threads>
+                    <<<blocks, threads, shared_memory_size>>>(std::move(fun));
+            else
+                launch_kernel_impl_::zero_extent_wrapper<num_threads, BlockSizeI, BlockSizeJ>
```

Comment:
do we gain a lot with it? has it compile time impact as we now compile to kernel launches?

Reply:
yes, it's a trade off. I have added TODO to investigate if it has positive impact.